README.rst
==========
Shortbread
==========


.. image:: https://img.shields.io/pypi/v/shortbread.svg
        :target: https://pypi.python.org/pypi/shortbread

.. image:: https://img.shields.io/travis/purplediane/shortbread.svg
        :target: https://travis-ci.org/purplediane/shortbread

.. image:: https://readthedocs.org/projects/shortbread/badge/?version=latest
        :target: https://shortbread.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Shortbread Recipe Calculator.


* Free software: MIT license
* Documentation: https://shortbread.readthedocs.io.


Features
--------

* Calculates and prints the ingredients for ``n`` dozen shortbread cookies.
* Prints a reminder of temperature and time for baking.
* Additional options exist for printing long instructions and a Chocolate Chip verion.
* Of course, the actual number of cookies and baking time depends upon how thick and large you make the cookies!
* See `Usage documentation`_ for hints and detailed instructions.

Credits
-------

This package created via the course `Creating and Distributing Python Packages`_ by Audrey and Danny Roy Greenfeld.

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _`Creating and Distributing Python Packages`: https://courses.twoscoopspress.com/
.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _`Usage documentation`: http://shortbread.readthedocs.io/en/latest/usage.html


HISTORY.rst
=======
History
=======

0.5.0 (2018-03-18)
------------------

* Add use of module ``inflect`` for better output of quantities.
* Make count optional so prints recipe for 1 dozen cookies.
* Actually test all options added in 0.4.0.
* Improve documentation for CLI options.
* Improve usage documentation.
* Eradicate last vestiges of Python 2.7.

0.4.1 (2018-03-17)
------------------

* Minor docs improvements
* Improve tests for new options

0.4.0 (2018-03-16)
------------------

* Add Chocolate chip option
* Add long instructions option

0.3.0 - 0.3.4 (2018-03-14)
--------------------------

* Final Base Version with CLI

0.2.0 (2018-03-11)
------------------

* Add Docs

0.1.0 (2018-03-10)
------------------

* First release on PyPI.


tests/test_shortbread.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `shortbread` package."""

import pytest

from click.testing import CliRunner

from shortbread import shortbread
from shortbread import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def check_one_doz_plain(result):
    """Helper function for output of one dozen cookie recipe."""
    assert 'Shortbread Cookies - makes approx. 1 doz.' in result.output
    assert '1 cup flour' in result.output
    assert '0.25 cups sugar' in result.output
    assert '0.5 cups salted butter' in result.output
    assert '0.25 teaspoons vanilla (optional)' in result.output


def check_four_doz_plain(result):
    """Helper function for output of four dozen cookie recipe."""
    assert 'Shortbread Cookies - makes approx. 4 doz.' in result.output
    assert '4 cups flour' in result.output
    assert '1 cup sugar' in result.output
    assert '2 cups salted butter' in result.output
    assert '1 teaspoon vanilla (optional)' in result.output


def test_cli_no_count():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    check_one_doz_plain(result)
    assert 'Bake cookies at 350 degrees for 10-12 minutes' in result.output


def test_cli_simple_one_doz():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    result = runner.invoke(cli.main, ['1'])
    assert result.exit_code == 0
    check_one_doz_plain(result)
    assert 'Bake cookies at 350 degrees for 10-12 minutes' in result.output


def test_cli_simple_four_doz():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    result = runner.invoke(cli.main, ['4'])
    assert result.exit_code == 0
    check_four_doz_plain(result)
    assert 'Bake cookies at 350 degrees for 10-12 minutes' in result.output


def test_cli_long():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    result = runner.invoke(cli.main, ['4', '-l'])
    assert result.exit_code == 0
    check_four_doz_plain(result)
    assert '- Preheat oven to 350 degrees F.' in result.output
    assert '- Bake 10 to 12 minutes at 350 degrees' in result.output


def test_cli_chocolate_one_doz():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    result = runner.invoke(cli.main, ['1', '-c'])
    assert result.exit_code == 0
    assert ('Chocolate Chip Shortbread Cookies - makes approx. 1 doz.'
            in result.output)
    assert '0.25 teaspoons baking powder' in result.output
    assert '0.25 cups small semisweet chocolate chips' in result.output
    assert result.exit_code == 0


def test_cli_chocolate_four_doz():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    result = runner.invoke(cli.main, ['4', '-c'])
    assert ('Chocolate Chip Shortbread Cookies - makes approx. 4 doz.'
            in result.output)
    assert '1 teaspoon baking powder' in result.output
    assert '1 cup small semisweet chocolate chips' in result.output
    assert result.exit_code == 0
    result = runner.invoke(cli.main, ['1', '-c'])


def test_cli_chocolate_long():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    result = runner.invoke(cli.main, ['4', '-lc'])
    assert result.exit_code == 0
    assert ('Chocolate Chip Shortbread Cookies - makes approx. 4 doz.'
            in result.output)
    assert '- Sift together baking powder and flour.' in result.output
    assert '- Stir in the chocolate chips.' in result.output
    assert '- Flatten cookie balls slightly.' in result.output
    assert ('- Bake 15 to 20 minutes at 350 degrees until golden brown.'
            in result.output)


def test_cli_help():
    """Test the CLI in the simplest version."""
    runner = CliRunner()
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    # assert 'Usage: shortbread [OPTIONS] COUNT' in help_result.output
    # Can't use that ^^^ because it says Usage: main here.
    assert '--help      Show this message and exit.' in help_result.output
    assert 'Script to print shortbread recipe for COUNT dozen cookies.' \
        in help_result.output
    assert '-l, --long  Print the longer version with directions.'  \
        in help_result.output
    assert '-c, --choc  Print the Chocolate Chip Shortbread version.'  \
        in help_result.output


def test_shortbread(capsys):
    """Test the output from shortbread"""
    # shortbread() only prints; returns nothing
    assert shortbread.shortbread(4) is None
    # Capture the result of the arepas.ingredients() function call.
    captured = capsys.readouterr()
    # If we check captured, we can see that the ingredients have been printed.
    assert 'Shortbread Cookies - makes approx. 4 doz.' in captured.out
    assert "4 cups flour" in captured.out
    assert "1 cup sugar" in captured.out
    assert "2 cups salted butter" in captured.out
    assert "Bake cookies at 350 degrees for 10-12 minutes" in captured.out


shortbread/shortbread.py
# -*- coding: utf-8 -*-

"""Main module."""
import inflect

p = inflect.engine()


def integerize(num, count):
    """Calculate and return integer value if result is integer"""
    calc = num * count
    calc = int(calc) if calc.is_integer() else calc
    return calc


def shortbread(count, long_list=False, choc=False):
    """Prints ingredients for making `count` dozen shortbread cookies."""
    title = 'Shortbread'
    if choc:
        title = 'Chocolate Chip Shortbread'
    title += ' Cookies - makes approx. {} doz.'.format(count)
    quarter = integerize(.25, count)
    half = integerize(.5, count)
    print(title)
    print('{} {} flour'.format(count, p.plural('cup', count)))
    print('{} {} sugar (may use brown sugar)'.format(quarter,
          p.plural('cup', quarter)))
    print('{} {} salted butter (NOT margarine), softened'.format(half,
          p.plural('cup', half)))
    print('{} {} vanilla (optional)'.format(quarter,
          p.plural('teaspoon', quarter)))
    if choc:
        print('{} {} baking powder'.format(quarter,
              p.plural('teaspoon', quarter)))
        print('{} {} small semisweet chocolate chips'.format(quarter,
              p.plural('cup', quarter)))
    light = 15 if choc else 10
    dark = 20 if choc else 12
    baking = '- Bake {} to {} minutes at 350 '.format(light, dark)
    baking += 'degrees until golden brown.'
    baking_hint = 'Bake cookies at 350 degrees for '
    baking_hint += '{}-{} minutes'.format(light, dark)

    if not long_list:
        print(baking_hint)
        return
    print()
    print('- Preheat oven to 350 degrees F.')
    print('- Cream the butter and sugar together well.')
    print('- Stir in optional vanilla.')
    if choc:
        print('- Sift together baking powder and flour.')
    print('- Mix in the flour until it makes a stiff dough.')
    print('- Add a bit more flour if dough is too soft.')
    if choc:
        print('- Stir in the chocolate chips.')
        print('- Scoop and roll into walnut-sized balls.')
    else:
        print('- Roll out 1/4 to 1/2 inch thick and cut into cookies.')
    print('- Place on baking sheet; line with parchment paper if desired.')
    if choc:
        print('- Flatten cookie balls slightly.')
    print(baking)
    print('When cool, enjoy your delicious cookies!')
